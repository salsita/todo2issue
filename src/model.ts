export interface Todo {
  filename: string
  line: number
  text: string
  tag: string
  issueNumber?: number
  author?: string
}

export interface Issue {
  todos: Todo[]
  issueNumber?: number
  assignees?: string[]
}


export const groupTodosByFile = (todos: Todo[]) => todos.reduce(
  (map, todo) => {
    let fileUpdates = map.get(todo.filename)
    if (fileUpdates) {
      fileUpdates.push(todo)
    } else {
      fileUpdates = [todo]
      map.set(todo.filename, fileUpdates)
    }
    return map
  },
  new Map<string, Todo[]>()
)
