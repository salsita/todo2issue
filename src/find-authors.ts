import { groupTodosByFile, Todo } from './model'
import { blameFile } from './git'

const range = (start: number, endInclusive: number) => new Array(endInclusive - start + 1).fill(null).map((_, index) => start + index)

export async function findAuthors (root: string, todos: Todo[], authorsByEmail: Record<string, string>, ignoreUnresolvedAuthors: boolean = false) {
  const todosByFile = groupTodosByFile(todos)
  const missingAuthorEmails = new Set<string>()
  for (const [file, fileTodos] of todosByFile.entries()) {
    const blameInfos = await blameFile(root, file, fileTodos.map(todo => todo.line))
    const blamesPerLine = new Map(
      blameInfos.flatMap(blameInfo =>
        range(blameInfo.lineStart, blameInfo.lineEnd).map(line => [line, blameInfo])
      )
    )
    todos.forEach(todo => {
      const blameInfo = blamesPerLine.get(todo.line)
      if(blameInfo && blameInfo.author.email !== 'not.committed.yet') {
        todo.author = authorsByEmail[blameInfo.author.email]
        if(!todo.author) {
          missingAuthorEmails.add(blameInfo.author.email)
        }
      }
    })
  }

  if(missingAuthorEmails.size > 0) {
    const message = `cannot resolve GitHub users from emails:\n\t${Array.from(missingAuthorEmails).join('\n\t')}`
    if(ignoreUnresolvedAuthors) {
      console.warn(message)
    } else {
      throw new Error(message)
    }
  }
}
