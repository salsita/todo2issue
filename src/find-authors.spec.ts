import { resolve } from 'path'
import { Todo } from './model'
import { findAuthors } from './find-authors'

describe('findAuthors', () => {
  it('finds authors', async () => {

    const root = resolve(__dirname, '../')
    const todos: Todo[] = [
      {
        line: 1,
        text: 'A',
        tag: 'TODO',
        filename: 'sample/src/index.ts'
      },
      {
        line: 4,
        text: 'B',
        tag: 'TODO',
        filename: 'sample/src/index.ts'
      },
      {
        line: 1,
        text: 'C',
        tag: 'TODO',
        filename: 'sample/src/helpers/helper.ts'
      },
      {
        line: 16,
        text: 'line comments',
        tag: 'TODO',
        filename: 'sample/src/index.ts'
      }
    ]
    await findAuthors(root, todos, {'jirist@salsitasoft.com': 'goce-cz'})
    expect(todos).toMatchSnapshot()
  })
})
