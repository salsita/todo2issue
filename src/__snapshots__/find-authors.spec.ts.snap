// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findAuthors finds authors 1`] = `
Array [
  Object {
    "author": "goce-cz",
    "filename": "sample/src/index.ts",
    "line": 1,
    "tag": "TODO",
    "text": "A",
  },
  Object {
    "author": "goce-cz",
    "filename": "sample/src/index.ts",
    "line": 4,
    "tag": "TODO",
    "text": "B",
  },
  Object {
    "author": "goce-cz",
    "filename": "sample/src/helpers/helper.ts",
    "line": 1,
    "tag": "TODO",
    "text": "C",
  },
  Object {
    "author": "goce-cz",
    "filename": "sample/src/index.ts",
    "line": 16,
    "tag": "TODO",
    "text": "line comments",
  },
]
`;
